manifests/init.pp

```diff
@@ -1,22 +1,83 @@
-# Class: sftp_jail
-# ===========================
+# @summary sftp_jail class
 #
-# Full description of class sftp_jail here.
+# @param chroot_base
+#   The directories path, which holds all jails.
 #
-# Parameters
-# ----------
+# @param sftp_users
+#   See `Sftp_jail::Sftp_Users` and `Sftp_jail::Sftp_User` types.
 #
-# * `sample parameter`
-#   Explanation of what this parameter affects and what it defaults to.
-#   e.g. "Specify one or more upstream ntp servers as an array."
+# @param manage_users
+#   Should this module manage the user resources?
+#
+# @param sub_dirs
+#   Default subdirectories to ensure in every users home. Can be overwritten for
+#   each user seperatly.
+#
+# @param merge_subdirs
+#   Merge sub_dirs with the default sub_dirs?
+#
+# @param admin_sub_dirs
+#   Default subdirectories to ensure in every admins home. Can be overwritten for
+#   each admin seperatly.
+#
+# @param password_authentication
+#   Password Authentication setting for SFTP users and Admins.
 #
 class sftp_jail (
-  Stdlib::Absolutepath $chroot_base = '/chroot',
+  Sftp_jail::Sftp_Users    $sftp_users              = {},
+  Stdlib::Absolutepath     $chroot_base             = '/chroot',
+  Boolean                  $manage_users            = false,
+  Sftp_jail::Sub_dirs      $sub_dirs                = [],
+  Boolean                  $merge_subdirs           = false,
+  Sftp_jail::Sub_dirs      $admin_sub_dirs          = [],
+  Sftp_jail::Pw_auth       $password_authentication = {},
 ) {
+  $sftp_admins = $sftp_users.filter |$k, $v| { $v['sftp_admin'] }
+
   file { $chroot_base:
     ensure => 'directory',
     owner  => 'root',
     group  => 'root',
     mode   => '0755',
   }
+
+  unless $sftp_admins.empty {
+    file { "${chroot_base}/home":
+      ensure  => 'directory',
+      owner   => 'root',
+      group   => 'root',
+      mode    => '0755',
+      require => File[$chroot_base],
+    }
+    $sftp_user_groups = ($sftp_users - $sftp_admins).filter |$k, $v| {
+      $v['ensure'] == 'present'
+    }.map |$k, $v| {
+      if $v['group'] {
+        $v['group']
+      } else {
+        $k
+      }
+    }
+  }
+
+  $sftp_users.each |$k, $v| {
+    if $v['sftp_admin'] {
+      $jail_base = $chroot_base
+      $sub_dirs = $admin_sub_dirs
+      $merge_subdirs = false
+    }
```

Comment:
Should there be an `else` for this?

Reply:
This variables are passed to sftp_jail::jail. The defaults of sftp_jail are just fine for normal users. Only SFTP admins need different values:
* They don't get the same set of subdirectories in there home.
* There Jail is different from the normal users (one directory higher, so they can access all users jails).
* Admins should not merge there subdirectory lists by default (can be overwritten for each user, it's just a different default for admins)
I don't know if this is "the right way", it is what I came up with. ;)

Reply:
I think that it would fail with strict variables. If you don't care about the setting's value, `$jail_base = undef` is probably the right thing to do. Then it should fall back to the default in define.

Reply:
I googled strict variables quickly. Makes somewhat sense, i'll change it.

Reply:
fixed: https://github.com/voxpupuli/puppet-sftp_jail/pull/69/commits/9fc6d1521f3a36222b6c1221c990b1b1cc964157
$sub_dirs and $merge_subdirs do exist in the class scope. I could:
```
$sub_dirs = $sub_dirs
```
to make it more obvious (?)